core/test_utils.py

```diff
@@ -0,0 +1,65 @@
+import unittest
+from unittest import mock
+import numpy as np
+from core import utils
+
+
+class TestUtils(unittest.TestCase):
+    @mock.patch("core.utils.np.random.normal", return_value=np.array([1, 0]))
+    def test_randomly_perturb_direction(self, mock_normal: mock.Mock):
+        new_direction = utils.randomly_perturb_direction(np.array([0, 1]), 1)
+        expected_direction = np.array([np.sqrt(0.5), np.sqrt(0.5)])
+        for new_val, expected_val in zip(new_direction, expected_direction):
```

Comment:
After parsing this, I think I understand that you're trying to check each index of the new and expected direction arrays for being almost equal? I personally find that this process adds more mental computation to the reader and makes it harder to check the correctness of this test. 

Since you're already using NumPy for this, I'm wondering if it'd be more readable and robust to use NumPy's testing method for this purpose, given by this function signature: `np.testing.assert_almost_equal(actual, desired, decimal=7, err_msg='', verbose=True)`

You can check out the docs here: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html

 This comment applies here and in the subsequent tests where you use this approach.

Reply:
This great! Much clearer. Thanks for finding this.